Feature/stm32 usart wide data

JvanDooren opened the pull request:
drivers: serial: stm32: add wide data support

Add wide data support to STM32 into
- interrupt driven code
- async code
- polling code
- DTS init code

Adds 9 bit datawidth support to serial driver.
Validated interrupt driven code on an STM32H743.

Signed-off-by: Jeroen van Dooren <jeroen.van.dooren@nobleo.nl>

erwango reviewed (changes requested):
Thanks for this change.

To ease the review process, I'd suggest to split changes into multiple elementary commits.
There are too much changes impacting the driver in a single commit to allow a clear assessment of the change and impact.
For instance:
- all changes that are not directly linked to WIDE_DATA introduction should be
split from the changes introducing WIDE_DATA
- formating changes should be done in a dedicated commit
- ...

Also, I suggest submitting the non STM32 related changes in a dedicated PR to get a broader audience.

JvanDooren commented:
> Thanks for this change.
> 
> To ease the review process, I'd suggest to split changes into multiple elementary commits. There are too much changes impacting the driver in a single commit to allow a clear assessment of the change and impact. For instance:
> 
>     * all changes that are not directly linked to WIDE_DATA introduction should be
>       split from the changes introducing WIDE_DATA
> 
>     * formating changes should be done in a dedicated commit
> 
>     * ...
> 
> 
> Also, I suggest submitting the non STM32 related changes in a dedicated PR to get a broader audience.

I guess you're addressing the dts binding change? I made that change because I saw we're slowly moving from runtime configuration to devicetree configuration and these were missing. Shall I split that off into a dedicated PR and keep the change here? Because this PR doesn't build without it.

erwango commented:
> I guess you're addressing the dts binding change? I made that change because I saw we're slowly moving from runtime configuration to devicetree configuration and these were missing. Shall I split that off into a dedicated PR and keep the change here? 

DTS indeed but also test changes (which are btw welcome so that we are able to verify functional status).

> Because this PR doesn't build without it.

Sure, but you can put them in a stand alone PR which commits are also included in current PR.
Same with tests btw.
Once dts and tests PR are accepted current PR could be rebased.

JvanDooren commented:
> > I guess you're addressing the dts binding change? I made that change because I saw we're slowly moving from runtime configuration to devicetree configuration and these were missing. Shall I split that off into a dedicated PR and keep the change here?
> 
> DTS indeed but also test changes (which are btw welcome so that we are able to verify functional status).
> 
> > Because this PR doesn't build without it.
> 
> Sure, but you can put them in a stand alone PR which commits are also included in current PR. Same with tests btw. Once dts and tests PR are accepted current PR could be rebased.

Added another PR for only the binding change: [binding](https://github.com/zephyrproject-rtos/zephyr/pull/54400)

erwango commented:
> Added another PR for only the binding change: https://github.com/zephyrproject-rtos/zephyr/pull/54400

Thanks.

Note that my [comment](https://github.com/zephyrproject-rtos/zephyr/pull/54354#pullrequestreview-1280812850) regarding the driver changes still stands.

JvanDooren commented:
> > Added another PR for only the binding change: #54400
> 
> Thanks.
> 
> Note that my [comment](https://github.com/zephyrproject-rtos/zephyr/pull/54354#pullrequestreview-1280812850) regarding the driver changes still stands.

Thanks.
About the formatting changes, all changes I did relate to WIDE support, and I formatted those touched parts with clang (actually the whole file should be formatted once because I had to revert a lot of lines changed).
What is then left to do?

erwango commented:
> > > Added another PR for only the binding change: #54400
> > 
> > 
> > Thanks.
> > Note that my [comment](https://github.com/zephyrproject-rtos/zephyr/pull/54354#pullrequestreview-1280812850) regarding the driver changes still stands.
> 
> Thanks. About the formatting changes, all changes I did relate to WIDE support, and I formatted those touched parts with clang (actually the whole file should be formatted once because I had to revert a lot of lines changed). What is then left to do?

To make it simple you impact code which is not linked to WIDE_DATA feature (such as changes to `uart_stm32_poll_out`).
This should be done in a specific commit (with a commit message which explains the reason of the change).
For instance, you're introducing `poll_out_fn` and `poll_in_f` type_def. Why is that required ? A dedicated commit with a proper justification is needed...
Then there will be a commit with WIDE_DATA specific bits (roughly all the code under `#ifdef CONFIG_UART_WIDE_DATA`)
Note that this is the minimum split. Depending of the nature of the changes more commits might be required.

JvanDooren committed:
drivers: serial: stm32: prevent stop/data-bits HW support code duplication

Preventing code duplication of macros checking for HW support on
stop bits and data-bits during runtime configuration.
Validated runtime configuration on an STM32H743.

Signed-off-by: Jeroen van Dooren <jeroen.van.dooren@nobleo.nl>

JvanDooren committed:
drivers: serial: stm32: rework code for max re-use in wide data support

Restructuring code for poll_in/poll_out/fifo_fill/fifo_read because for
wide data support, all code is identical except the calls to
LL_USART_{ReceiveData8/TransmitData8}.
This allows both implementations, 8 and 9 bit data-width to call a
visitor function, passing the either the 8 bit or 9 bit data-width
function pointer.

Signed-off-by: Jeroen van Dooren <jeroen.van.dooren@nobleo.nl>

JvanDooren committed:
drivers: serial: stm32: add wide data support

Add wide data support to STM32.
Validated interrupt driven code on an STM32H743.

Signed-off-by: Jeroen van Dooren <jeroen.van.dooren@nobleo.nl>

JvanDooren commented:
> > > > Added another PR for only the binding change: #54400
> > > 
> > > 
> > > Thanks.
> > > Note that my [comment](https://github.com/zephyrproject-rtos/zephyr/pull/54354#pullrequestreview-1280812850) regarding the driver changes still stands.
> > 
> > 
> > Thanks. About the formatting changes, all changes I did relate to WIDE support, and I formatted those touched parts with clang (actually the whole file should be formatted once because I had to revert a lot of lines changed). What is then left to do?
> 
> To make it simple: You impact code which is not linked to WIDE_DATA feature (such as changes to `uart_stm32_poll_out`), this should be done in a specific commit (with a commit message which explains the reason of the change). Also, you're introducing `poll_out_fn` and `poll_in_f` type_def. Why is that required ? A dedicated commit with a proper justification is needed... Then there will be a commit with WIDE_DATA specific bits (roughly all the code under `#ifdef CONFIG_UART_WIDE_DATA`) Note that this is the minimum split. Depending of the nature of the changes more commits might be required.

Ok, splitted the single commit into 4 separate ones.

erwango commented:
@JvanDooren I'll try to review next week

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -654,6 +675,29 @@ static int uart_stm32_err_check(const struct device *dev)
 	return err;
 }
 
+#ifdef CONFIG_UART_WIDE_DATA
```
Why isn't this section not with other poll functions ?

JvanDooren replied:
Moved

JvanDooren replied:
My bad, I missed a commit somehow

erwango reviewed (changes requested):
Here is a first round of review.

I suggest to use a clear and consistent naming convention to ease reader's understanding, and also to group related functions together.

Last, I'm not yet sure on the gain about using visitor function vs the added complexity. I'll take more time to review this part in deep.

JvanDooren commented on drivers/serial/uart_stm32.c:
```diff
@@ -282,19 +282,32 @@ static inline enum uart_config_parity uart_stm32_ll2cfg_parity(uint32_t parity)
 }
 
-static inline uint32_t uart_stm32_cfg2ll_stopbits(enum uart_config_stop_bits sb)
+static inline uint32_t uart_stm32_cfg2ll_stopbits(const struct uart_stm32_config *config,
```
Actually, this one is called only once.

erwango replied:
My bad, this comment actually applies to `uart_stm32_ll2cfg_databits`.

erwango replied:
Ok, I had a try and doesn't have any effect. I guess the compiler is doing the job efficiently..

JvanDooren commented on drivers/serial/uart_stm32.c:
```diff
@@ -675,6 +675,29 @@ static int uart_stm32_err_check(const struct device *dev)
+	*((uint16_t *)in) = LL_USART_ReceiveData9(config->usart);
+}
+
+static int uart_stm32_poll_in_u16(const struct device *dev, uint16_t *p_u16)
```
Fixed

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -675,6 +675,29 @@ static int uart_stm32_err_check(const struct device *dev)
+	LL_USART_TransmitData9(config->usart, *((uint16_t *)out));
+}
+
+static void uart_stm32_poll_in_u9(const struct uart_stm32_config *config, void *in)
+{
+	*((uint16_t *)in) = LL_USART_ReceiveData9(config->usart);
+}
+
+static int uart_stm32_poll_in_u16(const struct device *dev, uint16_t *p_u16)
```
I still think the function naming is confusing:
`uart_stm32_poll_in_u16/uart_stm32_poll_in` are 'API functions' while `uart_stm32_poll_in_u8/uart_stm32_poll_in_u9` are arguments to the visitor function. (same for poll_out ... ) of course.
I suggest not using `uart_stm32` prefix for arg functions (avoid `__` also which usage is restricted)

JvanDooren replied:
Agreed, will change

erwango reviewed (changes requested):
Thanks for this work.
Globally this is fine, but I'd like some changes on the function names to make it easier to understand.

Also, I'd like this new code to be compiled and tested. Please add matching test cases with one STM32 target to `tests/drivers/uart`

JvanDooren commented:
> Thanks for this work. Globally this is fine, but I'd like some changes on the function names to make it easier to understand.
> 
> Also, I'd like this new code to be compiled and tested. Please add matching test cases with one STM32 target to `tests/drivers/uart`

I'll add some tests

JvanDooren committed:
tests: drivers: uart: moving struct declaration

Moving struct in preparation of wide test.

Signed-off-by: Jeroen van Dooren <jeroen.van.dooren@nobleo.nl>

JvanDooren committed:
tests: drivers: uart: Adding wide data cfg tests

Extending configuration tests with wide data support.

Signed-off-by: Jeroen van Dooren <jeroen.van.dooren@nobleo.nl>

JvanDooren committed:
tests: drivers: uart: Resetting uart state

Resetting uart state before every testcase.

Signed-off-by: Jeroen van Dooren <jeroen.van.dooren@nobleo.nl>

JvanDooren committed:
tests: drivers: uart: Adding wide data async tests

Extending async tests with wide data support.

Signed-off-by: Jeroen van Dooren <jeroen.van.dooren@nobleo.nl>

JvanDooren commented:
@erwango @ABOSTM @FRASTM Any chance that this PR gets some focus on the short term?

erwango commented:
> @erwango @ABOSTM @FRASTM Any chance that this PR gets some focus on the short term?

Was on holidays. I'm back now, will try to review this week.

erwango commented:
@dcpleung Can you review the test part ?

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -679,45 +726,124 @@ static int uart_stm32_fifo_fill(const struct device *dev,
 	return num_rx;
 }
 
+static void uart_stm32_fifo_read_with_u8(const struct uart_stm32_config *config, void *rx_data,
```
Please apply the same naming convention than for poll functions

JvanDooren replied:
Fixed

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -1575,15 +1722,28 @@ static const struct uart_driver_api uart_stm32_driver_api = {
 	.irq_update = uart_stm32_irq_update,
 	.irq_callback_set = uart_stm32_irq_callback_set,
-#endif	/* CONFIG_UART_INTERRUPT_DRIVEN */
+#ifdef CONFIG_UART_WIDE_DATA
+	.fifo_fill_u16 = uart_stm32_fifo_fill_u16,
+	.fifo_read_u16 = uart_stm32_fifo_read_u16,
+#endif
```
Non blocking: It's only my preference here, but I feel this would be clearer if functions are grouped by purpose:
```
	.poll_in = uart_stm32_poll_in,
	.poll_out = uart_stm32_poll_out,
#ifdef CONFIG_UART_WIDE_DATA
	.poll_in_u16 = uart_stm32_poll_in_u16,
	.poll_out_u16 = uart_stm32_poll_out_u16,
#endif
	.err_check = uart_stm32_err_check,
#ifdef CONFIG_UART_USE_RUNTIME_CONFIGURE
	.configure = uart_stm32_configure,
	.config_get = uart_stm32_config_get,
#endif /* CONFIG_UART_USE_RUNTIME_CONFIGURE */
#ifdef CONFIG_UART_INTERRUPT_DRIVEN
	.fifo_fill = uart_stm32_fifo_fill,
	.fifo_read = uart_stm32_fifo_read,
#ifdef CONFIG_UART_WIDE_DATA
	.fifo_fill_u16 = uart_stm32_fifo_fill_u16,
	.fifo_read_u16 = uart_stm32_fifo_read_u16,
#endif
	.irq_tx_enable = uart_stm32_irq_tx_enable,
	.irq_tx_disable = uart_stm32_irq_tx_disable,
	.irq_tx_ready = uart_stm32_irq_tx_ready,
	.irq_tx_complete = uart_stm32_irq_tx_complete,
	.irq_rx_enable = uart_stm32_irq_rx_enable,
	.irq_rx_disable = uart_stm32_irq_rx_disable,
	.irq_rx_ready = uart_stm32_irq_rx_ready,
	.irq_err_enable = uart_stm32_irq_err_enable,
	.irq_err_disable = uart_stm32_irq_err_disable,
	.irq_is_pending = uart_stm32_irq_is_pending,
	.irq_update = uart_stm32_irq_update,
	.irq_callback_set = uart_stm32_irq_callback_set,
#endif	/* CONFIG_UART_INTERRUPT_DRIVEN */

```

JvanDooren replied:
Fixed

erwango reviewed (changes requested):
One naming change requested for consistency.
I still need to check tests

erwango commented on tests/drivers/uart/uart_async_api/testcase.yaml:
```diff
@@ -11,14 +11,28 @@ tests:
+    platform_allow: nucleo_h743zi
+    integration_platforms:
+      - nucleo_h743zi
+    extra_args: DTC_OVERLAY_FILE="boards/nucleo_h743zi.overlay"
+      OVERLAY_CONFIG="boards/nucleo_h743zi.conf"
```
superfluous

JvanDooren replied:
Removed

erwango commented on tests/drivers/uart/uart_async_api/testcase.yaml:
```diff
@@ -11,14 +11,28 @@ tests:
     harness_config:
       fixture: gpio_loopback
-    depends_on: gpio
+    depends_on: gpio uart
```
I don't think the uart keyword is required

JvanDooren replied:
Removed

erwango commented on tests/drivers/uart/uart_async_api/src/test_uart_async_wide.c:
```diff
@@ -0,0 +1,126 @@
+	k_sem_reset(&rx_wide_disabled);
+}
+
+#ifdef CONFIG_USERSPACE
```
I don't see real added value testing `CONFIG_USERSPACE` here, I'd prefer you add `uart_async_chain_read_wide` to test `uart_rx_buf_rsp_u16`.
And maybe even remove `uart_async_single_read_wide` if `uart_async_chain_read_wide` can test all API s at once.

erwango reviewed (changes requested):
Some suggestions to extend test coverage to whole _wide API, whole not duplicating the whole tests.
Another idea would be to rework existing tests to add a way to configure them in wide version at compile time.

Would be great to have @dcpleung to comment on these proposals.

JvanDooren commented on tests/drivers/uart/uart_basic_api/src/test_uart_config_wide.c:
Renamed

erwango replied:
I can see that this comment wasn't finally taken into account, or change wasn't pushed.
I'm approving anyway, as I think this PR could be merge into 3.4, but this is a required change IMO.

JvanDooren replied:
@erwango  Renamed anyway, somehow slipped through.

JvanDooren commented on tests/drivers/uart/uart_async_api/testcase.yaml:
```diff
@@ -12,6 +12,18 @@ tests:
+    extra_configs:
+      - CONFIG_UART_WIDE_DATA=y
+    arch_allow: arm
+    platform_allow: nucleo_h743zi
```
Added

JvanDooren commented on tests/drivers/uart/uart_async_api/testcase.yaml:
```diff
@@ -12,6 +12,18 @@ tests:
       fixture: gpio_loopback
     depends_on: gpio
+  drivers.uart.async_api.wide:
+    filter: CONFIG_UART_CONSOLE and CONFIG_SERIAL_SUPPORT_ASYNC and not CONFIG_UART_MCUX_LPUART
```
Added

JvanDooren commented on tests/drivers/uart/uart_basic_api/testcase.yaml:
```diff
@@ -5,6 +5,16 @@ tests:
+    platform_allow: nucleo_h743zi
+    integration_platforms:
+      - nucleo_h743zi
+    extra_args: DTC_OVERLAY_FILE="boards/nucleo_h743zi.overlay"
```
Removed

dcpleung reviewed (commented):
I also think it is a good idea to extend the current tests to support wide data. Although it will make reading the source harder, a bunch of typedef and macro aliases should work.

JvanDooren commented on drivers/serial/uart_stm32.c:
```diff
@@ -779,6 +803,47 @@ static int uart_stm32_fifo_read(const struct device *dev, uint8_t *rx_data, cons
 
+#ifdef CONFIG_UART_WIDE_DATA
+
+static void fifo_fill_with_u16(const struct uart_stm32_config *config,
```
Agreed

dcpleung replied:
How about `uart_stm32_fifo_fill_u16` and `uart_stm32_fifo_read_u16`?

JvanDooren replied:
Fixed

JvanDooren commented on drivers/serial/uart_stm32.c:
```diff
@@ -596,26 +596,50 @@ static void uart_stm32_poll_out_visitor(const struct device *dev, void *out, pol
 static void uart_stm32_poll_out(const struct device *dev, unsigned char c)
 {
-	uart_stm32_poll_out_visitor(dev, (void *)&c, uart_stm32_poll_out_u8);
+	uart_stm32_poll_out_visitor(dev, (void *)&c, poll_out_u8);
```
Moved

JvanDooren commented:
> I also think it is a good idea to extend the current tests to support wide data. Although it will make reading the source harder, a bunch of typedef and macro aliases should work.

The question is, will doubling all the existing async tests in a wide flavor test anything additionally? If they both test the internal operation of the usart, then this will not give any increased coverage, nor security. Isn't proving that a write_u9 and read_u9 work fine enough? Abort, double-buffering and timeout tests share behavior between wide and non-wide?

talih0 reviewed (changes requested):
For xmc4xxx changes, uart_xmc4xx_dma_xx_cb() is not dead code.
See the macro UART_DMA_CHANNEL_INIT() which sets .dma_callback

JvanDooren commented:
> UART_DMA_CHANNEL_INIT

Then why did the build fail, I didn't touch it before...

erwango commented on tests/drivers/uart/uart_basic_api/testcase.yaml:
```diff
@@ -7,6 +7,13 @@ tests:
+    extra_configs:
+      - CONFIG_UART_WIDE_DATA=y
+    tags: drivers uart
+    filter: ( CONFIG_UART_CONSOLE and CONFIG_SERIAL_SUPPORT_WIDE_DATA )
```
I guess fixture is still required here (same as async version)
```
    harness: ztest
    harness_config:
      fixture: gpio_loopback
    depends_on: gpio
```

JvanDooren replied:
Well, there's nothing to loopback here, it's just setting and getting configuration.

erwango replied:
Right, my bad. Approved for now.

talih0 commented on drivers/serial/uart_xmc4xxx.c:
```diff
@@ -688,6 +688,7 @@ static int uart_xmc4xxx_async_rx_enable(const struct device *dev, uint8_t *buf,
 	return dma_start(data->dma_rx.dma_dev, data->dma_rx.dma_channel);
 }
 
+#ifdef CONFIG_UART_ASYNC_API
```
The async functions are already guarded by CONFIG_UART_ASYNC_API

talih0 commented:
> > UART_DMA_CHANNEL_INIT
> 
> Then why did the build fail, I didn't touch it before... Most likely because those methods aren't between #ifdef CONFIG_UART_ASYNC_API?

The tests drivers.uart.async_api.rtt for xmc45_relax_kit and xmc47_relax_kit don't seem to excluded after
93b63df7627bdb0a2d04f8af9bd6d0181e807389

dcpleung commented:
> And now that we're at it, the API is ambiguous, in every u16 call it mentions a `size_t len`. Now what does that mean, does it mean sizeof(buf_u16), which is the raw size of the memory being pointed to, so can hold /2 nr of uint16's, or does it mean number of 16 bit values, so essentially sizeof(buf_u16)/sizeof(uint16_t) ?

That should be the number of data items in the buffer.

JvanDooren commented:
> > > UART_DMA_CHANNEL_INIT
> > 
> > 
> > Then why did the build fail, I didn't touch it before... Most likely because those methods aren't between #ifdef CONFIG_UART_ASYNC_API?
> 
> The tests drivers.uart.async_api.rtt for xmc45_relax_kit and xmc47_relax_kit don't seem to be excluded after [93b63df](https://github.com/zephyrproject-rtos/zephyr/commit/93b63df7627bdb0a2d04f8af9bd6d0181e807389)

It's as if the list doesn't work, as they are excluded but the notation syntax has indeed changed. How is it possible that this passed CI?

erwango commented:
> How is it possible that this passed CI? I can't find any closed PR given that title, was CI bypassed?

See https://github.com/zephyrproject-rtos/zephyr/actions/runs/4928890357/jobs/8807828501#step:8:788

JvanDooren commented:
> > > I also think it is a good idea to extend the current tests to support wide data. Although it will make reading the source harder, a bunch of typedef and macro aliases should work.
> > 
> > 
> > The question is, will doubling all the existing async tests in a wide flavor test anything additionally? If they both test the internal operation of the usart, then this will not give any increased coverage, nor security. Isn't proving that a write_u9 and read_u9 work fine enough? Abort, double-buffering and timeout tests share behavior between wide and non-wide?
> 
> The idea is to prove that everything still works correctly. There is no need to duplicate all the tests for wide data. The idea would be to create another test case (in `testcase.yaml`) with wide data enabled using the same source code. This would require changing the existing tests to support both... and that's where all the ifdef and macros come in. For example, if wide data is enabled, call `uart_rx_buf_rsp_u16` instead of `uart_rx_buf_rsp`. So you can use the same set of tests without duplication.

Done

erwango commented on tests/drivers/uart/uart_basic_api/src/test_uart_wide.c:
test_uart_config_wide

JvanDooren replied:
Renamed

JvanDooren commented:
Maybe some of you are on holiday, any chance it's going to make it for 3.4 RC1?

talih0 commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -30,14 +38,33 @@ static void init_test(void)
+		.data_bits = UART_CFG_DATA_BITS_9,
+		.flow_ctrl = UART_CFG_FLOW_CTRL_NONE
+	};
+	__ASSERT_NO_MSG(uart_configure(uart_dev, &uart_cfg) == 0);
```
Runtime config is optional and not supported by all async uart boards.

JvanDooren replied:
True, but devicetree config isn't supported at all yet, so how would I get tests to execute on a uart that's configured for 9 bits data?

JvanDooren replied:
Added an additional KConfig to enable runtime cfg

talih0 commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -91,28 +118,33 @@ static void *single_read_setup(void)
-	/* Check also if sending from read only memory (e.g. flash) works. */
-	static const uint8_t tx_buf[5] = "test\n";
+	DATA_TYPE rx_buf[10] = {0};
+	const uint8_t nr_of_rx_values = sizeof(rx_buf)/sizeof(DATA_TYPE);
```
You can use the ARRAY_SIZE() macro and probably just use it in the function to avoid declaring an extra const variable.

JvanDooren replied:
Fixed

talih0 commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -91,28 +118,33 @@ static void *single_read_setup(void)
+	static const DATA_TYPE tx_buf[5] = "test\n";
+#else
+	static const DATA_TYPE tx_buf[5] = {0x0074, 0x0065, 0x0073, 0x0074, 0x000A};
+#endif
```
I think you can just replace all of this by

static const DATA_TYPE tx_buf[5] = { 't', 'e', 's', 't', 0};

JvanDooren replied:
Fixed

talih0 commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -16,10 +28,6 @@ ZTEST_BMEM volatile bool failed_in_isr;
 	DEVICE_DT_GET(UART_NODE);
 
-static void read_abort_timeout(struct k_timer *timer);
-K_TIMER_DEFINE(read_abort_timer, read_abort_timeout, NULL);
```
These were added in a previous commit. Probably should be reviewed by the commit author @nordic-krch

JvanDooren replied:
Ok, the timer solution is a multi-threading disaster to be used in tests, which we should not do. Tests need to be deterministic.

JvanDooren commented:
WTF, why is a DMA driver failing on an ESP32. I don't see any relationship with my changes...

talih0 commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -680,11 +732,11 @@ ZTEST_USER(uart_async_chain_write, test_chained_write)
-ZTEST_BMEM uint8_t long_rx_buf2[1024];
-ZTEST_BMEM uint8_t long_tx_buf[1000];
-ZTEST_BMEM volatile uint8_t evt_num;
-ZTEST_BMEM size_t long_received[2];
```
I don't think you should change the array sizes. This is intentional and is used to test the driver's implementation. For example, it currently checks that the right number of UART_RX_RDY calls are made with this set up of arrays in the test_long_buffers test.

JvanDooren replied:
Fixed

JvanDooren commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -30,14 +38,33 @@ static void init_test(void)
+		.data_bits = UART_CFG_DATA_BITS_8,
+		.flow_ctrl = UART_CFG_FLOW_CTRL_NONE
+	};
+	__ASSERT_NO_MSG(uart_configure(uart_dev, &uart_cfg) == 0);
```
Maybe, but that also might not be supported, as runtime configuration existed at least in Zephyr 2.6, whereas compile (DTS) time didn't (for instance, the STM32 implementation didn't support it until I added it in this PR).
What might work is to add the CONFIG_UART_USE_RUNTIME_CONFIGURE to the filter?

talih0 replied:
I think that removing this #ifndef block would be enough. As I understand uart_configure() here is only needed for the wide test?

Extra filtering would remove the coverage from boards where this currently passes.

JvanDooren commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -93,7 +94,7 @@ ZTEST_USER(uart_async_single_read, test_single_read)
 
 	/* Check also if sending from read only memory (e.g. flash) works. */
-	static const uint8_t tx_buf[5] = "test";
+	static const uint8_t tx_buf[5] = "test\n";
```
True, but that memcpy down below was comparing a random last byte, because it wasn't filled in. It still works, but I didn't find it pretty.

dcpleung replied:
If you insist that something must be there, use `'\0'`.

heinwessels replied:
Changed to `\0`

JvanDooren commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -150,13 +177,16 @@ static void *multiple_rx_enable_setup(void)
+	static const DATA_TYPE tx_buf[] = "test";
+#else
+	static const DATA_TYPE tx_buf[] = {0x0074, 0x0065, 0x0073, 0x0074};
+#endif
```
Fixed, missed this one

JvanDooren commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -6,6 +6,18 @@
+#ifndef CONFIG_UART_WIDE_DATA
+#define DATA_TYPE uint8_t
+#define ENABLE_RX uart_rx_enable
+#define TX uart_tx
```
Fixed

erwango commented on tests/drivers/uart/uart_basic_api/src/test_uart_config_wide.c:
```diff
@@ -0,0 +1,121 @@
+
+#include "test_uart.h"
+
+#if DT_NODE_EXISTS(DT_NODELABEL(dut))
+#define UART_NODE DT_NODELABEL(dut)
+#else
+#define UART_NODE DT_CHOSEN(zephyr_console)
+#endif
```
If test can work on console, I wonder why we would also implement a dut version

JvanDooren replied:
Test can't work on console, because twister uses that same console for result reporting? At least twister execution breaks if I reconfigure that console to 9 bit data width.

erwango replied:
ok, that what I though. Then why the `else` ?

JvanDooren replied:
Because I didn't want to add all the overlays for all the boards on which this test can be executed. So, the test will build atm, but indeed might not execute. I can't test execution on those boards because I don't have them.
Same construct was already in zephyr/tests/drivers/uart/uart_async_api/src/test_uart.h, so I reused it here.

talih0 commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -427,35 +469,33 @@ static void *read_abort_setup(void)
 	zassert_equal(k_sem_take(&rx_disabled, K_MSEC(100)), 0,
 		      "RX_DISABLED timeout");
 	zassert_false(failed_in_isr, "Unexpected order of uart events");
-	zassert_not_equal(memcmp(tx_buf, rx_buf, 100), 0, "Buffers equal");
```
@JvanDooren, currently the test checks that receive was actually aborted by making sure that tx_buf and rx_buf don't have the same contents. With the proposed change it would only check that uart_rx_disable(uart_dev) doesn't return an error. There is nothing to confirm that the receiving actually stopped. This makes the test weaker IMO. 
Does this test fail on your hardware without the change?

JvanDooren replied:
It did fail, but somehow not anymore, reverted the change

talih0 commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -383,14 +426,20 @@ static void test_read_abort_callback(const struct device *dev,
+		if (tx_done_num == 2) {
+			err = uart_rx_disable(uart_dev);
+			failed_in_isr |= (err < 0);
+		}
```
Please see my comments below about changes to this test

JvanDooren replied:
Reverted

talih0 commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -607,8 +646,16 @@ ZTEST_USER(uart_async_timeout, test_forever_timeout)
+						     {0x004d, 0x0065, 0x0073, 0x0073,
+						      0x0061, 0x0067, 0x0065, 0x0020,
+						      0x0032}};
+#endif
```
nit: maybe this can be combined into one statement too?

JvanDooren replied:
Fixed

talih0 commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -6,6 +6,18 @@
 
+#ifndef CONFIG_UART_WIDE_DATA
+#define DATA_TYPE uint8_t
+#define UART_ENABLE_RX uart_rx_enable
```
nit: UART_RX_ENABLE

JvanDooren replied:
Fixed

talih0 commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -6,6 +6,18 @@
+#define DATA_TYPE uint16_t
+#define UART_ENABLE_RX uart_rx_enable_u16
+#define UART_TX uart_tx_u16
+#define UART_RX_RESP uart_rx_buf_rsp_u16
```
nit: UART_RX_BUF_RSP

JvanDooren replied:
Fixed

erwango commented on tests/drivers/uart/uart_async_api/testcase.yaml:
```diff
@@ -30,6 +30,18 @@ tests:
+    harness_config:
+      fixture: gpio_loopback
+    depends_on: gpio
+    extra_configs:
+      - CONFIG_UART_WIDE_DATA=y
+      - CONFIG_UART_USE_RUNTIME_CONFIGURE=y
```
What about using a "overlay-wide.conf" file with ` - OVERLAY_CONFIG="overlay-usb.conf"` here, see `drivers.uart.basic_api.cdc_acm` test variant. This would ease manual testing using west.

JvanDooren replied:
Fixed

erwango commented on tests/drivers/uart/uart_basic_api/testcase.yaml:
```diff
@@ -7,6 +7,14 @@ tests:
       - mps2_an385
+  drivers.uart.basic_api.wide:
+    extra_configs:
+      - CONFIG_UART_WIDE_DATA=y
```
Ditto about using a "overlay-wide.conf" file

JvanDooren replied:
Fixed

erwango commented on tests/drivers/uart/uart_basic_api/testcase.yaml:
```diff
@@ -7,6 +7,14 @@ tests:
+    tags: drivers uart
+    filter: CONFIG_UART_CONSOLE and CONFIG_SERIAL_SUPPORT_WIDE_DATA
+    integration_platforms:
+      - nucleo_h743zi
```
Please add `    harness: keyboard`

JvanDooren replied:
Fixed

talih0 commented:
> Yes, the configure is only needed for the wide-test, however, if wide tests are done with execution



> @talih0
> 
> > I think that removing this #ifndef block would be enough. As I understand uart_configure() here is only needed for the wide test?
> 
> Yes, the configure is only needed for the wide-test, however, if wide tests are done with execution, they leave the HW in the state where the UART peripheral is configured for 9 data bits. Unless the HW is rebooted between tests, the non-wide tests should fail because the UART peripheral hasn't reverted to its default 8 data bit state.

I always thought it's safe to assume each test run in it's own separate context. i.e. drivers.uart.async_api is built and flashed on the board, then drivers.uart.async_api.wide is built and flashed on the board. Wouldn't there be lots of problems if a previous test could affect the state of the current test?

JvanDooren commented:
> > Yes, the configure is only needed for the wide-test, however, if wide tests are done with execution
> 
> > @talih0
> > > I think that removing this #ifndef block would be enough. As I understand uart_configure() here is only needed for the wide test?
> > 
> > 
> > Yes, the configure is only needed for the wide-test, however, if wide tests are done with execution, they leave the HW in the state where the UART peripheral is configured for 9 data bits. Unless the HW is rebooted between tests, the non-wide tests should fail because the UART peripheral hasn't reverted to its default 8 data bit state.
> 
> I always thought it's safe to assume each test run in it's own separate context. i.e. drivers.uart.async_api is built and flashed on the board, then drivers.uart.async_api.wide is built and flashed on the board. Wouldn't there be lots of problems if a previous test could affect the state of the current test?

Ok, if that is the case, then I stand corrected. Only used configure() for wide tests.

talih0 commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -736,27 +774,27 @@ ZTEST_USER(uart_async_long_buf, test_long_buffers)
 	zassert_equal(k_sem_take(&tx_done, K_MSEC(200)), 0, "TX_DONE timeout");
 	zassert_equal(k_sem_take(&rx_rdy, K_MSEC(200)), 0, "RX_RDY timeout");
-	zassert_equal(long_received[0], 500, "Wrong number of bytes received.");
-	zassert_equal(memcmp(long_tx_buf, long_rx_buf, 500),
```
Do we still need to make changes to the callback in UART_RX_RDY in this test? i.e. could we keep callback unchanged and change this part to:
zassert_equal(long_received[0], 500 * sizeof(DATA_TYPE), "Wrong number of bytes received.");
zassert_equal(memcmp(long_tx_buf, long_rx_buf, 500  * sizeof(DATA_TYPE) ),

JvanDooren replied:
Reverted all my changes

JvanDooren commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -607,8 +640,8 @@ ZTEST_USER(uart_async_timeout, test_forever_timeout)
-
-ZTEST_DMEM uint8_t chained_write_tx_bufs[2][10] = {"Message 1", "Message 2"};
+ZTEST_DMEM DATA_TYPE chained_write_tx_bufs[2][10] = {{'M', 'e', 's', 's', 'a', 'g', 'e', '1'},
+						     {'M', 'e', 's', 's', 'a', 'g', 'e', '2'}};
```
Fixed

JvanDooren commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -6,6 +6,18 @@
+#define DATA_TYPE uint16_t
+#define UART_RX_ENABLE uart_rx_enable_u16
+#define UART_TX uart_tx_u16
+#define UART_RX_BUF_RESP uart_rx_buf_rsp_u16
```
Fixed

JvanDooren commented on tests/drivers/uart/uart_async_api/src/test_uart_async.c:
```diff
@@ -23,6 +23,13 @@ K_TIMER_DEFINE(read_abort_timer, read_abort_timeout, NULL);
+	k_sem_reset(&tx_aborted);
+	k_sem_reset(&rx_rdy);
+	k_sem_reset(&rx_buf_released);
+	k_sem_reset(&rx_disabled);
```
Removed

talih0 replied:
They were not removed in read_abort_setup(), mutliple_rx_enable_setup(), multiple_rx_enable_setup().

heinwessels replied:
This has been removed.

JvanDooren commented:
C'mon guys, how many times will we break the code-base? dma_mcux_lpc.c is broken, last week it was something else.

JvanDooren commented on drivers/serial/uart_stm32.c:
```diff
@@ -282,19 +282,32 @@ static inline enum uart_config_parity uart_stm32_ll2cfg_parity(uint32_t parity)
+#if HAS_LPUART_1
+		if (IS_LPUART_INSTANCE(config->usart)) {
+			/* return the default */
+			return LL_USART_STOPBITS_2;
```
Hmm, good point, unclear why I didn't spot this...

heinwessels replied:
Changed

erwango replied:
@heinwessels Please consider https://github.com/zephyrproject-rtos/zephyr/pull/59406, I don't know which one will be ready for merge first, but they're colliding on some areas.

heinwessels replied:
If I get this PR ready in the today or tomorrow, would it be possible to merge this PR first? We don't really have time allocated to this PR, en solving big merge conflicts might halt working on it again. Obivously depends on the extent. I do believe there is only one comment I still need to address on this PR as well, but of course it also need to get approvals.

JvanDooren commented:
Ok, gentlemen, I've been given an ultimatum by my client, who's been paying for this work for 5 months already. I'll abandon this PR in 2 weeks, after which we'll continue on our fork. The process of this PR has been a deception till now, and there have been too many review rounds, each time finding new issues that needed to be fixed. When I see that [this](https://github.com/zephyrproject-rtos/zephyr/pull/57847) can be done in three weeks, where the whole async test is rewritten, I get the feeling that we're measuring with different criteria.

When some-one is up for rebasing this PR and testing again, please do so but I'm not spending any more of my time nor my client's money.

erwango commented:
> Based on your commitment to continue the test changes in another PR, I'm ok to drop the testing related commit from this PR and speed up the merging process for driver changes. (Note that it will have to wait V3.5.0 window opening though, as there is nothing I can do on this front).

Ok, I understand your point.
What do you think about my proposal ?

erwango commented:
Approved on my side.
@talih0 @dcpleung What's your take on testing part of this PR ?
Is it ok for you ? 
If not I propose this is split in another PR and at least the drivers part get merged

dcpleung reviewed (commented):
The test changes look good.

talih0 commented:
@erwango, @dcpleung, I'm not blocking the PR if you think it's good enough.

erwango commented:
> I'm not blocking the PR if you think it's good as is.

@talih0 thanks. Would you mind to dismissing your review, as it still appears as "change requested" and will block the merge.

@JvanDooren are you still on board ?

JvanDooren commented:
Well, I stopped responding because I have other urgent things on my mind, getting to a stable system before end of september, when we launch the product.

erwango commented:
DNM until author is back (@JvanDooren don't hesitate to remove label when needed)

benner commented:
Nice initiative @JvanDooren. Can't wait for it to be merged.

JvanDooren commented:
> Rebased, and fixed small conflict to `test_uart_async.c`. I ran the async tests on the nucleo_h743zi and the async tests pass. Will wait for original author to come back from vacation to verify if that's all that's needed, because he mentioned that there's a lot of work to be done to get catch up to `main`. If this is still the case might drop this PR again. Simply trying to nudge it forward.
> 
> ```
> scripts/twister -c -T tests/drivers/uart/uart_async_api --platform nucleo_h743zi --device-testing --device-serial /dev/ttyACM0 --fixture gpio_loopback
> ```
> 
> ```
> INFO    - 4/5 nucleo_h743zi             tests/drivers/uart/uart_async_api/drivers.uart.async_api PASSED (device 11.958s)
> INFO    - 5/5 nucleo_h743zi             tests/drivers/uart/uart_async_api/drivers.uart.async_api.wide PASSED (device 12.373s)
> ```

Nope, the tests were basically the main culprit. So should be fine if all tests pass.

JvanDooren commented:
> DNM until author is back (@JvanDooren don't hesitate to remove label when needed)

Doesn't seem I'm authorized to edit labels?

heinwessels commented:
@erwango As far as I can see all comments are addressed. So now we only have to wait for approvals.

erwango commented:
> @erwango As far as I can see all comments are addressed. So now we only have to wait for approvals.

For the tests part, I'd like @dcpleung to approve

dcpleung reviewed (approved):
Changes in tests LGTM.
